src/dailyai/services/websocket_transport_service.py

```diff
@@ -0,0 +1,121 @@
+import asyncio
+import time
+from typing import AsyncGenerator, List
+import websockets
+
+from dailyai.pipeline.frame_processor import FrameProcessor
+from dailyai.pipeline.frames import AudioFrame, ControlFrame, EndFrame, Frame, TTSEndFrame, TTSStartFrame, TextFrame
+from dailyai.pipeline.pipeline import Pipeline
+from dailyai.serializers.protobuf_serializer import ProtobufFrameSerializer
+from dailyai.services.base_transport_service import BaseTransportService
+
+
+class WebSocketFrameProcessor(FrameProcessor):
+    """This FrameProcessor filters and mutates frames before they're sent over the websocket.
+    This is necessary to aggregate audio frames into sizes that are cleanly playable by the client"""
+
+    def __init__(
+            self,
+            audio_frame_size=16000,
+            sendable_frames: List[Frame] | None = None):
+        super().__init__()
+        self._audio_frame_size = audio_frame_size
+        self._sendable_frames = sendable_frames or [TextFrame, AudioFrame]
+        self._audio_buffer = bytes()
+        self._in_tts_audio = False
+
+    async def process_frame(self, frame: Frame) -> AsyncGenerator[Frame, None]:
+        if isinstance(frame, TTSStartFrame):
+            self._in_tts_audio = True
+        elif isinstance(frame, AudioFrame):
+            if self._in_tts_audio:
+                self._audio_buffer += frame.data
+                while len(self._audio_buffer) >= self._audio_frame_size:
+                    yield AudioFrame(self._audio_buffer[:self._audio_frame_size])
+                    self._audio_buffer = self._audio_buffer[self._audio_frame_size:]
+        elif isinstance(frame, TTSEndFrame):
+            self._in_tts_audio = False
+            if self._audio_buffer:
+                yield AudioFrame(self._audio_buffer)
+                self._audio_buffer = bytes()
+        elif type(frame) in self._sendable_frames or isinstance(frame, ControlFrame):
+            yield frame
+
+
+class WebsocketTransport(BaseTransportService):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._sample_width = kwargs.get("sample_width", 2)
+        self._n_channels = kwargs.get("n_channels", 1)
+        self._port = kwargs.get("port", 8765)
+        self._host = kwargs.get("host", "localhost")
+        self._audio_frame_size = kwargs.get("audio_frame_size", 16000)
+        self._sendable_frames = kwargs.get(
+            "sendable_frames", [
+                TextFrame, AudioFrame, TTSEndFrame, TTSStartFrame])
+        self._serializer = kwargs.get("serializer", ProtobufFrameSerializer())
+
+        if self._camera_enabled:
+            raise ValueError(
+                "Camera is not supported in WebsocketTransportService")
+
+        if self._speaker_enabled:
```

Comment:
same here.

Reply:
it seems `_speaker_buffer_pending` is not used. so maybe these too blocks are not needed? seems a transport should need to care about speaker or camera.

Reply:
sorry, this comes from the base class....

Reply:
but i believe the comment about transport not caring about cameras or microphone still applies... This might be related to your other comment of separating transport from services. I think it makes total sense.